Support opening and saving files with legacy encodings

tomopumipumi opened the pull request:
## Summary

Addresses #16965

This PR adds support for **opening and saving** files with legacy encodings (non-UTF-8).
Previously, Zed failed to open files encoded in Shift-JIS, EUC-JP, Big5, etc., displaying a "Could not open file" error screen. This PR implements automatic encoding detection upon opening and ensures the original encoding is preserved when saving.

## Implementation Details

1.  **Worktree (Loading)**:
    * Updated `load_file` to use `chardetng` for automatic encoding detection.
    * Files are decoded to UTF-8 internal strings for editing, while preserving the detected `Encoding` metadata.
2.  **Language / Buffer**:
    * Added an `encoding` field to the `Buffer` struct to store the detected encoding.
3.  **Worktree (Saving)**:
    * Updated `write_file` to accept the stored encoding.
    * **Performance Optimization**:
        * **UTF-8 Path**: Uses the existing optimized `fs.save` (streaming chunks directly from Rope), ensuring no performance regression for the vast majority of files.
        * **Legacy Encoding Path**: Implemented a fallback that converts the Rope to a contiguous `String/Bytes` in memory, re-encodes it to the target format (e.g., Shift-JIS), and writes it to disk.
        * *Note*: This fallback involves memory allocation, but it is necessary to support legacy encodings without refactoring the `fs` crate's streaming interfaces.

## Changes

- `crates/worktree`:
    - Add dependencies: `encoding_rs`, `chardetng`.
    - Update `load_file` to detect encoding and decode content.
    - Update `write_file` to handle re-encoding on save.
- `crates/language`: Add `encoding` field and accessors to `Buffer`.
- `crates/project`: Pass encoding information between Worktree and Buffer.
- `crates/vim`: Update `:w` command to use the new `write_file` signature.

## Verification

I validated this manually using a Rust script to generate test files with various encodings.

**Results:**

* ✅ **Success (Opened & Saved correctly):**
    * **Japanese:** `Shift-JIS` (CP932), `EUC-JP`, `ISO-2022-JP`
    * **Chinese:** `Big5` (Traditional), `GBK/GB2312` (Simplified)
    * **Western/Unicode:** `Windows-1252` (CP1252), `UTF-16LE`, `UTF-16BE`
* ⚠️ **limitations (Detection accuracy):**
    * Some specific encodings like `KOI8-R` or generic `Latin1` (ISO-8859-1) may partially display replacement characters (`?`) depending on the file content length. This is a known limitation of the heuristic detection library (`chardetng`) rather than the saving logic.


Release Notes:

- Added support for opening and saving files with legacy encodings (Shift-JIS, Big5, etc.)

tomopumipumi committed:
feat: Support opening and saving files with legacy encodings (Shift-JIS,
etc.)

SomeoneToIgnore commented:
We actually used to have quite a promising PR https://github.com/zed-industries/zed/pull/36497 for a similar thing, but will leave for @ConradIrwin to check on both and decide.

tomopumipumi commented:
Thanks for the information. I'm happy to wait for Conrad to take a look.

ConradIrwin commented:
@tomopumipumi Amazing, thank you for this! I love that it's much simpler than the previous approaches we were trying. 

In testing out this PR against https://github.com/zed-industries/encodings-tests, I noticed that if you open `ISO_2022-JP` or `UTF16-LE` and then save the file without changing, it writes different bytes to disk.

I want to be sure that Zed is (to the greatest extent possible) not going to silently corrupt files, so I think we should make sure that this doesn't happen.

I'd also like (but probably a separate PR) a status bar indicator that shows when the current file is not UTF-8 to avoid surprises (and then there's the even more scope creep of being able to select and change encodings; but they should definitely be follow ups).

tomopumipumi committed:
feat: prioritize BOM-based encoding detection and ensure persistence

- Perform immediate encoding detection if a BOM is present during file
  load in worktree.
- Add `has_bom` flag to `Buffer` to track original BOM presence.
- Ensure the original BOM is re-inserted when saving the buffer.
- Fix byte-for-byte mismatch issues by strictly following the detected
  BOM.

tomopumipumi committed:
Update crates/worktree/src/worktree.rs

Co-authored-by: CrazyboyQCD <53971641+CrazyboyQCD@users.noreply.github.com>

tomopumipumi commented on crates/worktree/src/worktree.rs:
```diff
@@ -1464,7 +1481,49 @@ impl LocalWorktree {
+                    bytes
+                } else {
+                    cow.into_owned()
+                };
```
@CrazyboyQCD Thanks for the review! Avoiding the allocation in the else branch is a great improvement. I've applied your suggestion.
Regarding writing the BOM and content separately, I think I'll stick to the single buffer approach for now to keep the implementation simple.

tomopumipumi committed:
Merge remote-tracking branch 'upstream/main' into feat/detect-file-encoding

ConradIrwin committed:
cargo machete

zed-zippy[bot] committed:
Autofix

CrazyboyQCD commented:
@tomopumipumi 
I forgot that `UTF-16LE` and `UTF-16BE` in `encoding_rs` are decode-only encodings, current solution is [problematic](https://github.com/hsivonen/encoding_rs/issues/104), they should be done separately like [this](https://github.com/EuclidDivisionLemma/zed/blob/4330e5ff1439be5c0b2a88203b456379526abf84/crates/encodings/src/lib.rs#L98-L113).

tomopumipumi commented:
​@CrazyboyQCD Thanks for the explanation! I understand the issue now.
​I plan to open a separate PR to fix this UTF-16 saving logic and also implement the detection heuristic Conrad suggested. I think it's better to address these logic improvements together before working on the UI.
